### PR TITLE
Fix importing some (broken) beatmaps looping infinitely

### PR DESCRIPTION
**Pull request by peppy**
Resolves ppy/osu#1988.

Not sure on the path we want to take forward when handling these kinds of situations. An alternative to this implementation is to hard-fail on import or load when a beatmap does not match known standards.

### CODE REVIEW

File: osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs [hunk 1]
```diff
@@ -42,6 +43,9 @@ protected override void ParseBeatmap(StreamReader stream, Beatmap beatmap)
 
             ParseContent(stream);
 
+            // objects may be out of order if user is a cunt.
```
Comment (smoogipoo) [inline osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs]: no

File: osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs [hunk 2]
```diff
@@ -326,7 +330,7 @@ private void handleTimingPoints(string line)
                 beatmap.ControlPointInfo.DifficultyPoints.Add(new DifficultyControlPoint
                 {
                     Time = time,
-                    SpeedMultiplier = speedMultiplier
+                    SpeedMultiplier = MathHelper.Clamp(speedMultiplier, 0.01, 100)
```
Comment (smoogipoo) [inline osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs]: I really don't like this. Let's make this 0.001 at least. A simple map I tested has 0.05, and I'm sure there are harsher maps that go even lower than that.

File: osu.Game/Rulesets/Objects/Legacy/Osu/ConvertHitObjectParser.cs [hunk 3]
```diff
@@ -29,7 +30,7 @@ protected override HitObject CreateSlider(Vector2 position, bool newCombo, List<
                 Position = position,
                 NewCombo = newCombo,
                 ControlPoints = controlPoints,
-                Distance = length,
+                Distance = Math.Max(0, length),
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Objects/Legacy/Osu/ConvertHitObjectParser.cs]: This is not going to stand in the future. We'll need an issue thread for this.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs (hunk 2 above)
Comment (MillhioreF) [reply osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs]: Maps are already clamped to these values in stable (and in fact are clamped between 0.1 and 10.0 in standard). They display as lower or higher in the editor if changed in notepad but gameplay remains within the clamped values.

Comment (smoogipoo) [reply osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs]: Can you point me to the code in osu-stable that does this? I don't remember seeing such a clamping outside of the editor...

Comment (smoogipoo) [reply osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs]: Oh, controlpoint, I see...

Comment (peppy) [reply osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs]: Thank fuck for that.

For reference:

File: osu.Game/Beatmaps/ControlPoints/TimingControlPoint.cs [hunk 4]
```diff
@@ -15,6 +16,12 @@ public class TimingControlPoint : ControlPoint
         /// <summary>
         /// The beat length at this control point.
         /// </summary>
-        public double BeatLength = 1000;
+        public double BeatLength
+        {
+            get => beatLength;
+            set => beatLength = MathHelper.Clamp(value, 100, 60000);
```
Comment (smoogipoo) [inline osu.Game/Beatmaps/ControlPoints/TimingControlPoint.cs]: Range not adjusted as stated

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**